### PR TITLE
Added llama-index integration

### PR DESCRIPTION
**Pull request by jverre**
## Details
Llama Index integration, includes documentation updates with sample code snippets

### CODE REVIEW

File: sdks/python/src/opik/integrations/llama_index/callback.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+        event_starts_to_ignore = (
+            event_starts_to_ignore if event_starts_to_ignore else []
+        )
+        event_ends_to_ignore = event_ends_to_ignore if event_ends_to_ignore else []
+        super().__init__(
+            event_starts_to_ignore=event_starts_to_ignore,
+            event_ends_to_ignore=event_ends_to_ignore,
+        )
+
+        self._opik_client = Opik()
```
Comment (alexkuzmik) [inline sdks/python/src/opik/integrations/llama_index/callback.py]: `opik_client.get_client_cached()`

File: sdks/python/src/opik/integrations/llama_index/callback.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+        self._opik_trace: Optional[Trace] = None
+
+        self._map_event_id_to_span: Dict[str, Any] = {}
```
Comment (alexkuzmik) [inline sdks/python/src/opik/integrations/llama_index/callback.py]: if it's event id to span, then it should be `Dict[str, span.Span]`

File: sdks/python/src/opik/integrations/llama_index/callback.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+        # When a trace finishes, we first get the last event output
+        last_event = self._get_last_event(trace_map)
+        last_event_output = self._map_event_id_to_output.get(last_event, None)
+
+        # And then end the trace with the optional output
+        if self._opik_trace:
+            self._opik_trace.end(output=last_event_output)
+            self._opik_trace = None
+
+    def _get_span_output_from_event(
```
Comment (alexkuzmik) [inline sdks/python/src/opik/integrations/llama_index/callback.py]: `_get_span_input` and `_get_span_output` can be moved to a separate module as functions.

File: sdks/python/src/opik/integrations/llama_index/callback.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
+from typing import Optional, Dict, List, Any
+
+from llama_index.core.callbacks.base_handler import BaseCallbackHandler
+from llama_index.core.callbacks.schema import CBEventType
```
Comment (alexkuzmik) [inline sdks/python/src/opik/integrations/llama_index/callback.py]: don't import classes, functions and other variables (but you can still import names from `typing`). Import modules and then access what you need like that
```python
from llama_index.core.callbacks import base_handler
class LlamaIndexCallbackHandler(base_handler.BaseCallbackHandler):
```

You can use `import ... as ...` to get a more descriptive module name if without its parents it looks too generic.

File: sdks/python/src/opik/integrations/llama_index/callback.py [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+            event_ends_to_ignore=event_ends_to_ignore,
+        )
+
+        self._opik_client = Opik()
+        self._opik_trace: Optional[Trace] = None
+
+        self._map_event_id_to_span: Dict[str, Any] = {}
+        self._map_event_id_to_output: Dict[str, Any] = {}
+
+    def _create_trace(self, trace_id: Optional[str] = None) -> Trace:
```
Comment (alexkuzmik) [inline sdks/python/src/opik/integrations/llama_index/callback.py]: Let's remove default value for trace_id (if it's not the overriden function from the base class), it's not used anywhere in the code.
Unused default values on different levels of the code are traps.

And let's maybe rename argument name to `trace_name`.

File: sdks/python/src/opik/integrations/llama_index/callback.py [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+
+                return {"output": output}
+
+        if payload_copy:
+            return {"output": payload_copy}
+        else:
+            return None
+
+    def _get_span_input_from_events(
+        self, event_type: CBEventType, payload: Optional[Dict[str, Any]] = None
```
Comment (alexkuzmik) [inline sdks/python/src/opik/integrations/llama_index/callback.py]: no need for default value for payload.

File: sdks/python/src/opik/integrations/llama_index/callback.py [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+        last_event = self._get_last_event(trace_map)
+        last_event_output = self._map_event_id_to_output.get(last_event, None)
+
+        # And then end the trace with the optional output
+        if self._opik_trace:
+            self._opik_trace.end(output=last_event_output)
+            self._opik_trace = None
+
+    def _get_span_output_from_event(
+        self, event_type: CBEventType, payload: Optional[Dict[str, Any]] = None
```
Comment (alexkuzmik) [inline sdks/python/src/opik/integrations/llama_index/callback.py]: no need for default value for payload

File: sdks/python/src/opik/integrations/llama_index/callback.py [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+    ) -> None:
+        # Get the span output from the event and store it so we can use it if needed
+        # when finishing the trace
+        span_output = self._get_span_output_from_event(event_type, payload)
+        if event_id:
+            self._map_event_id_to_output[event_id] = span_output
+
+            # Log the output to the span with the matching id
+            span = self._map_event_id_to_span[event_id]
+            if span:
```
Comment (alexkuzmik) [inline sdks/python/src/opik/integrations/llama_index/callback.py]: When is it possible? Based on the usage of `_map_event_id_to_span` - you always put not None values (spans) into dict.
And if `event_id` is not in the dict at all - the previous line will raise KeyError.

**Review by alexkuzmik — CHANGES_REQUESTED** [review]

File: sdks/python/src/opik/integrations/llama_index/callback.py (hunk 8 above)
Comment (jverre) [reply sdks/python/src/opik/integrations/llama_index/callback.py]: Makes sense, updated it to not fail if the span is missing for whatever reason

**Review by alexkuzmik — APPROVED** [review]

**Pull request merged**